src/main/scala/chisel3/util/SparseVec.scala

```diff
@@ -0,0 +1,230 @@
+    case index =>
+      if (!duplicates.add(index))
+        nonUniqueMapping = true
+      index.toString -> DataMirror.internal.chiselTypeClone(gen)
+  })
+
+  // Throw a runtime exception if there is a non-unique mapping.
+  // TODO: Improve this error message.
+  if (nonUniqueMapping) {
+    throw new ChiselException("non-unique mapping in SparseVec")
+  }
+
+  // A zeroValue is the value of the vector at index zero.  The zeroValue is
+  // important because it will set the out-of-bounds behavior of the SparseVec.
+  private val zeroValue: Option[A] = elements.get("0").asInstanceOf[Some[A]]
+
+  // Determine if a default value needs to exist.  This should only exist if
+  // there are fewer map items than the size of the Vec.  (As this is a
+  // SparseVec, this is expected to be a Some.)
+  private val defaultValue: Option[A] = mapping match {
+    case full if mapping.size == size => None
+    case _                            => Some(WireInit(gen, DontCare))
+  }
+
+  /** Read a value from a [[SparseVec]] using one of several possible lookup
+    * typs. The returned value is read-only.
+    *
+    * @param addr the address of the value to read from the vec
+    * @param lookupType the type of lookup, e.g., binary, one-hot, or when-based
+    * @param sourceinfo implicit source locator information
+    * @return a read-only value from the specified address
+    * @throws ChiselException if the returned value is written to
+    */
+  def apply(addr: UInt, lookupType: SparseVec.Type = SparseVec.Binary)(implicit sourceinfo: SourceInfo): A = {
+    val result: A = lookupType match {
+      // Generate a lookup using a decoder.
+      case d: SparseVec.Decoder =>
+        // The number of bits required to represent all addresses of the vector.
+        val addrWidth = log2Up(size)
+
+        // The address where user-provided data starts.  If there is no default
+        // value needed, then start user data at address zero.
+        val baseEncodedAddress = defaultValue match {
+          case Some(_) => 1
+          case None    => 0
+        }
+
+        // The total number of output addresses that need to be encoded.
+        val encodedSize = elements.size + baseEncodedAddress + {
+          zeroValue match {
+            case Some(_) => 1
+            case None    => 0
+          }
+        }
+
+        val legalBitPats = mapping.zipWithIndex.map {
+          case (index, i) =>
+            BitPat(index.U(addrWidth.W)) -> BitPat(d.encoding(i + baseEncodedAddress, encodedSize))
+        } ++ {
+          // If out-of-bounds behavior is possible, then add BitPats for all
+          // out-of-bounds accesses.
+          zeroValue match {
+            case None => Seq.empty
+            case Some(_) =>
+              (size until BigInt(addrWidth).pow(2).toInt).map {
+                case i =>
+                  BitPat(i.U(addrWidth.W)) -> BitPat(d.encoding(baseEncodedAddress, encodedSize))
+              }
+          }
+        }
```

Comment:
Similarly the user may want to opt-in to dont care on these.

We also may want to check QoR of doing this this way vs. using a mux in the same way we do on IfElse below--but we can do that evaluation later.

Reply:
User can opt-in to indeterminate values if they want to.